File: java/src/main/java/com/qmates/kata/extractinterface/StripePaymentGateway.java
package com.qmates.kata.extractinterface;

import java.util.Map;

/**
 * Concrete payment gateway that talks to Stripe over HTTP.
 *
 * <p>Notice the shape of the problem: this is a single, heavyweight, I/O-bound
 * class. It needs an API key, builds a real {@link HttpClient}, and posts to a
 * live endpoint. There is no abstraction between {@link OrderProcessor} and
 * this class — {@code OrderProcessor} names this concrete type directly.
 */
public class StripePaymentGateway implements ChargePaymentGateway {

  private final HttpClient http = new HttpClient();
  private final String baseUrl = "https://api.stripe.com/v1";
  private final String apiKey;

  public StripePaymentGateway(String apiKey) {
    if (apiKey == null || apiKey.isBlank()) {
      throw new IllegalArgumentException("StripePaymentGateway requires an API key.");
    }
    this.apiKey = apiKey;
  }

  @Override
  public ChargeResult charge(long amountCents, String paymentToken) {
    Map<String, Object> payload =
        Map.of(
            "amount", amountCents,
            "currency", "usd",
            "source", paymentToken,
            "key", apiKey);

    Map<String, Object> response = http.post(baseUrl + "/charges", payload);
    return ChargeResult.succeeded((String) response.get("id"));
  }

  /** Issues a refund for a previously successful charge. */
  public void refund(String chargeId) {
    http.post(baseUrl + "/refunds", Map.of("charge", chargeId));
  }
}

File: typescript/src/01-extract-interface/stripe-payment-gateway.ts
import { HttpClient } from "./http-client.js";
import type { Cents } from "../domain.js";
import type ChargePaymentGateway from "./charge-payment-gateway";

export interface ChargeResult {
  success: boolean;
  /** Provider-side identifier for the charge, when successful. */
  chargeId?: string;
  /** Human-readable failure reason, when unsuccessful. */
  declineReason?: string;
}

/**
 * Concrete payment gateway that talks to Stripe over HTTP.
 *
 * Notice the shape of the problem: this is a single, heavyweight, I/O-bound
 * class. It needs an API key, builds a real `HttpClient`, and posts to a live
 * endpoint. There is no abstraction between `OrderProcessor` and this class —
 * `OrderProcessor` names this concrete type directly.
 */
export class StripePaymentGateway implements ChargePaymentGateway {
  private readonly http = new HttpClient();
  private readonly baseUrl = "https://api.stripe.com/v1";

  constructor(private readonly apiKey: string) {
    if (!apiKey) {
      throw new Error("StripePaymentGateway requires an API key.");
    }
  }

  async charge(amountCents: Cents, paymentToken: string): Promise<ChargeResult> {
    const response = (await this.http.post(`${this.baseUrl}/charges`, {
      amount: amountCents,
      currency: "usd",
      source: paymentToken,
      key: this.apiKey,
    })) as { id: string };

    return { success: true, chargeId: response.id };
  }

  /** Issues a refund for a previously successful charge. */
  async refund(chargeId: string): Promise<void> {
    await this.http.post(`${this.baseUrl}/refunds`, { charge: chargeId });
  }
}
